### PR TITLE
Mask vs sequence

### PR DESCRIPTION
**Pull request by evgeniiaraz**
**Proposed changes**:
- store sequence length instead of mask; compute mask when needed;

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/utils/tensorflow/model_data.py [hunk 1]
```diff
@@ -160,6 +160,20 @@ def add_mask(self, key: Text, from_key: Text):
...
 
+    def add_lengths(self, key: Text, from_key: Text):
+        if not self.data.get(from_key):
+            return
+
+        self.data[key] = []
+
+        for data in self.data[from_key]:
+            if data.size > 0:
+                # explicitly add last dimension to mask
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:171]: the comment does not seem to fit

File: rasa/utils/tensorflow/model_data.py [hunk 2]
```diff
@@ -160,6 +160,20 @@ def add_mask(self, key: Text, from_key: Text):
                 self.data[key].append(mask)
                 break
 
+    def add_lengths(self, key: Text, from_key: Text):
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:163]: please add return type and also a doc string

File: rasa/nlu/classifiers/diet_classifier.py [hunk 3]
```diff
@@ -1352,13 +1354,17 @@ def _calculate_entity_loss(
 
         return loss, f1
 
+    def compute_mask(self, sequence_lengths):
```
Comment (tabergma) [inline rasa/nlu/classifiers/diet_classifier.py:1357]: Please add types. Also I think this method can be private.

File: rasa/nlu/classifiers/diet_classifier.py [hunk 4]
```diff
@@ -1352,13 +1354,17 @@ def _calculate_entity_loss(
...
+        mask = tf.sequence_mask(sequence_lengths, dtype=tf.float32)
+        return tf.expand_dims(mask, -1)
```
Comment (tabergma) [inline rasa/nlu/classifiers/diet_classifier.py:1359]: Can you please add a comment, why we do `tf.expand_dims(mask, -1)`?

File: rasa/utils/tensorflow/model_data.py (hunk 2 above)
Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:163]: Not sure, but I think the `add_mask` method is not used anymore. We might want to remove it.

Comment (evgeniiaraz) [reply rasa/utils/tensorflow/model_data.py:147]: there isn't a return type because it doesn't return anything :) but I have added the docstring

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:147]: If there is no return value you should annotate it with `-> None`

**Comment by evgeniiaraz** [comment]
I haven't seen performance improvements F1-accuracy-wise (it's the same), and I wouldn't expect that because we're essentially doing the same thing. Would need to check memory / time!
I think I've addressed previous comments.

File: rasa/nlu/classifiers/diet_classifier.py [hunk 5]
```diff
@@ -1255,10 +1251,12 @@ def _create_sequence(
 
     def _create_all_labels(self) -> Tuple[tf.Tensor, tf.Tensor]:
         all_label_ids = self.tf_label_data[LABEL_IDS][0]
+
+        label_lengths = tf.cast(self.tf_label_data[LABEL_SEQ_LENGTH][0], dtype=tf.int32)
```
Comment (tabergma) [inline rasa/nlu/classifiers/diet_classifier.py:1255]: Should we add a helper method for this? Something like `sequence_length_for`? This logic is repeated all over the place.

File: rasa/nlu/classifiers/diet_classifier.py [hunk 6]
```diff
@@ -1352,13 +1350,19 @@ def _calculate_entity_loss(
 
         return loss, f1
 
+    def _compute_mask(self, sequence_lengths: tf.Tensor) -> tf.Tensor:
```
Comment (tabergma) [inline rasa/nlu/classifiers/diet_classifier.py:1353]: Can be static.

File: rasa/utils/tensorflow/model_data.py [hunk 7]
```diff
@@ -144,20 +144,19 @@ def add_features(self, key: Text, features: List[np.ndarray]):
         # update number of examples
         self.num_examples = self.number_of_examples()
 
-    def add_mask(self, key: Text, from_key: Text):
-        """Calculate mask for given key and put it under specified key."""
-
+    def add_lengths(self, key: Text, from_key: Text) -> None:
+        """
+        Adds np.array of lengths of sequences to data under given key
+        """
```
Comment (tabergma) [inline rasa/utils/tensorflow/model_data.py:150]:
```suggestion
        """Adds np.array of lengths of sequences to data under given key."""
```

**Comment by evgeniiaraz** [comment]
Timewise it is the same as the previous way of doing it.

**Pull request merged**